src/main/java/net/dries007/tfc/common/blocks/devices/BloomeryBlock.java

```diff
@@ -0,0 +1,306 @@
+            .matchHorizontal(origin.above(), stoneMatcher, 1);
+
+        BLOOMERY_BASE[2] = new MultiBlock() //west
+            .match(origin, center)
+            .match(origin.below(), stoneMatcher)
+            .match(origin.west(), state -> state.is(TFCBlocks.BLOOMERY.get()))
+            .matchEachDirection(origin, stoneMatcher, new Direction[]{Direction.NORTH, Direction.EAST, Direction.SOUTH}, 1)
+            .matchEachDirection(origin.west(), stoneMatcher, NORTH_SOUTH_DOWN, 1)
+            .matchHorizontal(origin.above(), stoneMatcher, 1);
+
+        BLOOMERY_BASE[3] = new MultiBlock() //east
+            .match(origin, center)
+            .match(origin.below(), stoneMatcher)
+            .match(origin.east(), state -> state.is(TFCBlocks.BLOOMERY.get()))
+            .matchEachDirection(origin, stoneMatcher, new Direction[]{Direction.NORTH, Direction.WEST, Direction.SOUTH}, 1)
+            .matchEachDirection(origin.east(), stoneMatcher, NORTH_SOUTH_DOWN, 1)
+            .matchHorizontal(origin.above(), stoneMatcher, 1);
+
+        BLOOMERY_CHIMNEY = new MultiBlock()
+            .match(origin, insideChimney)
+            .matchHorizontal(origin, stoneMatcher, 1);
+
+        // Gate center is the bloomery gate block
+        GATE_Z = new MultiBlock()
+            .match(origin, state -> state.is(TFCBlocks.BLOOMERY.get()) || state.isAir())
+            .matchEachDirection(origin, stoneMatcher, new Direction[]{Direction.WEST, Direction.EAST, Direction.UP, Direction.DOWN}, 1);
+
+        GATE_X = new MultiBlock()
+            .match(origin, state -> state.is(TFCBlocks.BLOOMERY.get()) || state.isAir())
+            .matchEachDirection(origin, stoneMatcher, new Direction[]{Direction.NORTH, Direction.SOUTH, Direction.UP, Direction.DOWN}, 1);
+    }
+
+    //centerPos should be the internal block of the bloomery
+    public static int getChimneyLevels(Level level, BlockPos centerPos)
+    {
+        for (int i = 1; i < 4; i++)
+        {
+            BlockPos center = centerPos.above(i);
+            if (!BLOOMERY_CHIMNEY.test(level, center))
+            {
+                return i - 1;
+            }
+        }
+        // Maximum levels
+        return 3;
+    }
+
+    public static boolean canGateStayInPlace(Level level, BlockPos pos, Direction.Axis axis)
+    {
+        if (axis == Direction.Axis.X)
+        {
+            return GATE_X.test(level, pos);
+        }
+        else
+        {
+            return GATE_Z.test(level, pos);
+        }
+    }
+
+    //if Directions don't have int indices any more, then...
+    public static boolean isFormed(Level level, BlockPos centerPos, Direction facing)
+    {
+        return switch (facing)
+            {
+                case NORTH -> BLOOMERY_BASE[0].test(level, centerPos);
+                case SOUTH -> BLOOMERY_BASE[1].test(level, centerPos);
+                case WEST -> BLOOMERY_BASE[2].test(level, centerPos);
+                case EAST -> BLOOMERY_BASE[3].test(level, centerPos);
```

Comment:
if you're going to do it like an array, you may as well use Direction#get2dDataValue here and make this one line